### ISSUE TITLE
E0603:  Module is private

### ISSUE THREAD

**Issue by David-OConnor** [intellij-rust/intellij-rust#2586]
It appears that this plugin, when used with PyCharm, is flagging all file-based modules as private with error E0603. It does not flag modules defined in the file being used. This red warning is displayed when importing the module.

Eg: main.rs has 
`mod mymod;`

Another file has
`use mymod;`
or `use mymod::{MyStruct};`

mymod (after use) has the zigzag with E0603 warning.

**Comment by Undin** [comment]
@David-OConnor 
Could you specify your plugin version and provide more extended code example?
I can't reproduce it with the following code (looks like it fits the description):
```rust
//- main.rs
mod mymod;
mod mymod2;
```
```rust
//- mymod.rs
pub struct MyStruct;
```
```rust
//- mymod2.rs
use mymod;
```

**Comment by David-OConnor** [comment]
0.2.0.2096-181

``` // -main.rs
mod transforms;
mod render;
```

``` // transforms.rs
pub fn view_matrix(  // ...
```

``` // render.rs
use transforms;  // Red squiggles over transforms with Module `Transforms` is private [E0603]
// ...
let view = transforms::view_matrix( //..  Red squiggles over traansforms with same error as above.
```

**Comment by aertoria** [comment]
Same error occur to me today. Follow up on this thread

**Comment by jasonish** [comment]
I'm getting this as well. I have a layout like:

./src/lib.rs
```
pub mod codec
```

./src/code.rs
```
/// codec source code
```

./src/bin/app.rs
```
use mycreatename::codec; // XXX Intellij tells me codec is private
```

**Comment by jasonish** [comment]
> I'm getting this as well. I have a layout like:
> 
> ./src/lib.rs
> 
> pub mod codec
> ./src/code.rs
> 
> /// codec source code
> ./src/bin/app.rs
> 
> use mycreatename::codec; // XXX Intellij tells me codec is private

Update on this. My crate had a main.rs and a lib.rs. I filed this issue while in the process of moving some of the commands from main.rs into src/bin. When I was done with that, and got rid of the main.rs, these modules are no longer detected as private.

**Comment by David-OConnor** [comment]
I also have both main.rs and lib.rs

**Comment by arik-so** [comment]
I'm getting the same error.

**Comment by alanhkarp** [comment]
I just started getting this error after a big merge.  It's only an annoyance, since the code compiles and runs, but I sure would like a fix.

**Comment by RobertBrown3** [comment]
I am getting this error when attempting to access a local module that *never* had a main.rs. It is a library with a single test function in lib.rs.

I am getting the "module is private" message despite the fact that everything in the module is declared public. There are no private submodules, just some structs, enums, and functions.

Has anyone released a fix for this?

**Comment by kluen** [comment]
Hi, I'm not sure if the error I'm getting is exactly the same use case, but it looks like it is related. I have the following files in my project:

`Cargo.toml`:
```
[package]
name = "example"
version = "0.1.0"
edition = "2018"

[dependencies]
piston2d-graphics = "0.30"
piston_window = "0.89"
```

`src/main.rs`:
```
use piston_window::*;

mod foo;

fn main() {
}
```
`src/foo.rs`
```
use graphics;
```


This compiles without errors with Cargo stable.

When using IntelliJ, the `use graphics;` declaration in `foo.rs` shows the error

> Module `graphics` is private [E0603]

However, if I remove `use piston_window::*;` in `main.rs`, the error in `foo.rs` disappears.

Hope this helps.

**Comment by alanhkarp** [comment]
Thanks for the hint, but what if main.rs uses something from piston_window?  That's my problem.  For example, main.rs has

>mod foo;
mod bar;
use foo::qux;
fn main() {
    qux();
}

and bar.rs has

>use foo:qux();


The bogus error is that module foo is private.  Unlike you, I don't have a use statement that I can remove.

I really think my problem has something to do with having multiple modules with main() functions.

**Comment by kluen** [comment]
@alanhkarp Yes, it seems to be a different error. I cannot reproduce your error however. Can you provide a full minimal example including all files? Might be more helpful to the devs here.

**Comment by alanhkarp** [comment]
Unfortunately, I'm under some time pressure for my next deliverable.  My
quick try at reproducing the error didn't work.  I'll keep your note in my
active list, but it may be a while before I get the time I need.

Here are some clues.

   - The problem started after a merge that added two more files with
   main() and added a main() to lib.rs.
   - Some modules show as private in one source file but not another.
   - Changing Cargo.toml to specify a single [[bin]] reduces the number of
   modules shown as private.

I was able to do some experiments.  If comment out all but one [[bin]] from
my Cargo.toml and comment out all of lib.rs, which has a main(), the
spurious error goes away.  For each additional [[bin]] in the Cargo.toml,
more of the modules show up as private in the IntelliJ editor.  I have not
tried to see if conditional compilation instead of commenting out the
[[bin]] has the same effect because I don't know how to do conditional
compilation :(.

--------------
Alan Karp

**Comment by alanhkarp** [comment]
One correction.  I just looked for main() in lib.rs, and it's not there.

--------------
Alan Karp

**Comment by vlad20012** [comment]
Now we know the reason of this problem, see #3757
